### ISSUE TITLE
Should `not_null<>` provide `reset()`, `release()`? Should they be SFINAE-transparent?

### ISSUE THREAD

**Issue by mbeutel** [gsl-lite/gsl-lite#207]
`not_null<T>` aims to be a transparent replacement for the underlying pointer-like type `T`. Does that mean we should also provide "forwarders" for typical smart-pointer member functions such as `release()` and `reset()`, similar to `get()` which we already provide?

Another question is whether we want `not_null<>` to be SFINAE-transparent too, i.e. if calling `.release()` on a `not_null<std::shared_ptr<>>` would be a substitution failure. Then we'd have to do gymnastics like

```c++
    template < class U = T >
    decltype( std::declval<U>().release() )
    release() { ... }
```

**Comment by petamas** [comment]
`reset()` and `release()` are designed to set the underlying pointer to null, so I think we shouldn't provide them.

**Comment by mbeutel** [comment]
Counter-argument: unary `reset()` can also be used to swap out the pointer for another, which is fine. We would assert that its argument is not `nullptr` (and `= delete` the `std::nullptr_t` overload).

Nullary `reset()` and `release()` would indeed leave behind a `nullptr`, but so does `gsl::as_nullable( std::move( p ) ).release()`. One could argue that `release()` permits explicitly moving the underlying pointer, and moving from `not_null<>` is something we support. Similarly, nullary `reset()` is the C++98 way of saying `discard( p )`, with `discard()` defined as

```c++
    // Discards a resource held by a handle object.
    // Modeled after Rust's `drop()`.
template< class T >
constexpr void discard( T& arg ) noexcept
{
    auto _ = std::move(arg);
    (void) _; // suppress compiler warning about unused variable
}
template< class T >
constexpr void discard( T const & arg ) = delete;
```

**Comment by mbeutel** [comment]
Proposed implementation in https://github.com/mbeutel/gsl-lite/commit/0994cfb4ede586d5ee9299795d4763f6be19416e. Please discuss.

**Comment by petamas** [comment]
I'll be AFK for a few days, but I'd really like to argue about it. Is it possible to delay this until after 0.36?

**Comment by mbeutel** [comment]
> I'll be AFK for a few days, but I'd really like to argue about it.

So would I :)

> Is it possible to delay this until after 0.36?

`reset()` and `release()`, absolutely. But would you be ok with just renaming (and thereby rescoping) `gsl_CONFIG_NOT_NULL_TRANSPARENT_GET` to `gsl_CONFIG_TRANSPARENT_NOT_NULL` for 0.36? Even if we ultimately don't do `reset()`/`release()` I think there are more opportunities for "forwarding" which are less controversial, so `gsl_CONFIG_TRANSPARENT_NOT_NULL` could cover these too.

**Comment by petamas** [comment]
I'm okay with the rename

**Comment by mbeutel** [comment]
After thinking about this a bit more, I'm inclined to do without `reset()` and `release()`.

However, I would like to retain the SFINAE for `not_null<>::get()` and the `operator []` forwarding from the referenced commit (so we can support `not_null<std::unique_ptr<T[]>>`).

**Comment by petamas** [comment]
Hi @mbeutel,

my "few days" turned out to be more than 9 months, I'm really sorry about that.

I agree with your revised stand on `reset()` and `release()`. My reason is that in my mind, the "moved-out-of" state is separate from the "containing nullptr" state, and while nullary `reset()` and `release()` are making it nullptr, while moving from creates a "moved-out-of" state (which is implemented as containing nullptr, but that's an implementation detail).

I see your argument about thinking of `release()` as a move-like operation, but I prefer to explicitly see the move (like in your example: `gsl::as_nullable(std::move(p)).release()`). If we really-really wanted to have `release()`, I think we should make it rvalue-only.

About `reset()`: unary version is fine (although I'd prefer to use `=make_not_null(...)`, making it unnecessary), but I don't think `drop()` would be very C++-like. (Not unless not using the variable after dropping it was enforced by the compiler, like I assume it does in Rust.)

In conclusion: I still think we shouldn't have these functions. For me, transparency means anything that can be done with a `not_null<T>` can be done with plain old `T` (allowing the implementation of `not_null` as an alias), but not in the opposite direction: things you can do with a `T` but not with a `not_null<T>` are fine. "Storing nullptr" is one of those things, and I believe `reset()` and `release()` belong with it too.

I'll react to your other modifications in a separate post in a few minutes.

**Comment by petamas** [comment]
SFINAE-ing `get()`: I don't see the value of it, as I don't think it would produce a more readable build errors, although I haven't tried it yet. Also, I see a few cons:
- makes `get()`'s return value independent of element_type, (this may or may not be a good thing, depending on you PoV.)
- SFINAE makes the build slower, as I learnt in one of my last PRs here
- it complicates the signature of the function, and `not_null` is already plenty unreadable to normal people
What are your pros for the SFINAE version?

`operator []`: I like the idea, but see my caveats about SFINAE above; also, if we _do_ SFINAE, then we should SFINAE the other two operators too (`*` and `->`, like in `unique_ptr`, where they're only defined for the non-array version)

Comments: I like them, they should be included in the next version.

**Comment by mbeutel** [comment]
> my "few days" turned out to be more than 9 months, I'm really sorry about that.

Not a problem at all. I was sidetracked by other projects as well and didn't have much time for library work myself. I appreciate your comments and your PRs.

I agree with your reasoning and think it would be best to not add any of the functions discussed in this issue, perhaps with the exception of `operator []`. But I'll address that when I have an actual need for it.
